### DISCUSSION TITLE
📋 [RFC] Unify API about picker's custom render

### DISCUSSION THREAD

**Discussion by kiner-tang** [ant-design/ant-design#41303] [RFCs - archive]
# 📋 RFC

## Summary

- Use `cellRender` instead of `dateRender` and `monthCellRender`.
- Support both of `cellRender` when the picker property's value is one of  `quarter`、`year`、`week`、`time`.

## Motivation

As  [#41255](https://github.com/ant-design/ant-design/issues/41255) mentioned. 

The current picker component has different apis for customizing render options cells based on different types. For example, render using dateRender when picker="date" and monthCellRender when picker="month".

However, some types do not support custom rendering, such as picker="quarter" or picker="year".

In order to facilitate unified management and user use, we need to unify the custom rendering api for all types of pickers.

### CellRenderInfo

```typescript
type CellRenderInfo<DateType = Dayjs> = {
  originNode: React.ReactNode,
  today: DateType,
  range?: 'start' | 'end',
  type: PickerMode,
};
```

| **Property** | **Description**                           | **Type**                                                     | **Default** |
| ------------ | ----------------------------------------- | ------------------------------------------------------------ | ----------- |
| `cellRender` | Render custom option cell in picker panel | `<DateType>(currentDate: DateType, info: CellRenderInfo) => ReactNode` | -           |

## Basic Example

```tsx
// cellRender
const cellRender = (current, info) => {
  const { originNode } = info;
  switch (type) {
    case "year":
      // will override cell wrapper node
      return React.cloneElement(<div className="customWrapper">{originNode}</div>, {}, <div style={{ color: 'red' }}>{current.year()}</div>);
    case "month":
      return <div style={{ color: 'green' }}>{current.month()}</div>;
    case "quarter":
    case "week":
    case "time":
      // ...
  }
}

return (
	<>
    <Picker<Moment>
      picker="year"
      cellRender={cellRender}
    />

    <Picker<Moment>
      picker="month"
      cellRender={cellRender}
    />
  </>
);
```

## Detailed

- Refactor  rc-picker.

- Mark properties such as `dateRender` and `monthCellRender` as `deprecated` in rc-picker.

- Mark properties such as `dateCellRender` , `monthCellRender`, `dateFullCellRender` and   `monthFullCellRender` as `deprecated` in ant-design.

- Update document about `cellRender`.

**Comment by zombieJ** [comment]
Can we?

```tsx
type CellRenderInfo = {
  today: DayJS,
  range: 'start' | 'end',
  type: xxx,
  Wrapper: ReactComponent,
};
type CellRender = (current: DayJS, info: CellRenderInfo) => React.ReactNode;
```

**Reply by zombieJ** [reply]
个人觉得这个 API 的 wrapper 设计不好

**Comment by zombieJ** [comment]
Or like this:
```tsx
type CellRenderInfo = {
  originNode: React.ReactNode,
  today: DayJS,
  range: 'start' | 'end',
  type: xxx,
};
type CellRender = (current: DayJS, info: CellRenderInfo) => React.ReactNode;
```

Customize with children:
```tsx
cellRender={(_, { originNode }) => {
  return React.cloneElement(originNode, {}, 'Hello World');
}}
```
